[GTK][WPE][Skia] Avoid texture copying for accelerated buffers if possible

nikolaszimmermann opened the pull request:
#### b131d74c1eed28d2d729f044ee72f8581e28e706
<pre>
[GTK][WPE][Skia] Avoid texture copying for accelerated buffers if possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=280066">https://bugs.webkit.org/show_bug.cgi?id=280066</a>

Reviewed by Miguel Gomez.

The AcceleratedBuffers nowadays wrap a BitmapTexture. If the
BitmapTexture covers the whole tile texture size, we can swap the tile
texture with the texture stored in the AcceleratedBuffer and save a
texture copy.

Covered by existing tests.

* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::BitmapTexture):
(WebCore::BitmapTexture::swapTexture):
(WebCore::BitmapTexture::reset):
(WebCore::BitmapTexture::updateContents):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):

Canonical link: <a href="https://commits.webkit.org/284250@main">https://commits.webkit.org/284250@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a220ab16f9d165eabda72dcbac93f4ab64d904a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21463 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72872 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19947 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70919 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55992 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19766 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54836 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13274 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71867 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44031 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59408 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35305 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40696 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16835 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18309 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62646 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17182 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74567 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12775 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16431 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62324 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12814 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62362 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15283 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10322 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3933 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43993 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45070 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46265 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44809 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -91,6 +91,22 @@ BitmapTexture::BitmapTexture(const IntSize& size, OptionSet<Flags> flags)
+    std::swap(m_textureFormat, other.m_textureFormat);
+    m_shouldClear = false;
+
+    // Swapping textures is only valid if there are no additional GL objects attached.
```
In theory we could make this a "bool trySwapTexture(BitmapTexture&)" method, and return false if any of the conditions in the assertions is not fulfilled -- however we're not reaching any of those at present. I documented this in the form of assertions, and kept the logic simple.

magomez replied:
On the compositor side, BitmapTextures are used for CoordinatedBackingStore tiles, but they are used as well when rendering TextureMapperLayers into intermediate surfaces. In that case it's when bindAsSurface() gets called and BitmaptTexture has its fbo created. It may also create the depth and stencil buffers if the DepthBuffer flag is present.

The textures used for CoordinatedBackingStore never request the DepthBuffer flag, so we're never getting a BitmapTexture with a depth and stencil buffer from the pool. But we may get one with an existent fbo.

When implementing this, we need to take into account that in the previous scenario, the texture has been left attached to the fbo. If someone renders to that fbo, the result will go the texture, even if it was swapped and doesn't belong anymore to this BitmapTexture.  To avoid that, we need to attach the new texture to the existent fbo, so everything stays as it was. The failing texture on the gtk bot could be related to this.

I'm not 100% sure, but I think m_flags should not be swapped. We're swapping the internal texture, not the BitmapTexture properties. Size should be the same, doesn't need to be swapped. TextureFormat is always RGBA in both cases, as we don't support any other format.

Then for pixelFormat: with the current code, we always use RGBA for both accelerated an unaccelerated buffers, so we don't need to swap anything. When we add the hybrid mode, we will still use RGBA always.
But we're going to add the change to use BGRA when using the cpu backend only, so we need to handle that case. But that case won't use this swapTexture call because the buffer is not backedByOpenGL(), so I think we don't need to swap for that case either.

nikolaszimmermann committed:
[GTK][WPE][Skia] Avoid texture copying for accelerated buffers if possible
https://bugs.webkit.org/show_bug.cgi?id=280066

Reviewed by Miguel Gomez.

The AcceleratedBuffers nowadays wrap a BitmapTexture. If the
BitmapTexture covers the whole tile texture size, we can swap the tile
texture with the texture stored in the AcceleratedBuffer and save a
texture copy.

Covered by existing tests.

* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::BitmapTexture):
(WebCore::BitmapTexture::swapTexture):
(WebCore::BitmapTexture::reset):
(WebCore::BitmapTexture::updateContents):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):

Canonical link: https://commits.webkit.org/284250@main
